plot.go

```diff
@@ -262,34 +272,96 @@ func (plot *Plot) drawAxesToScreen(screen tcell.Screen) {
 ) {
-	tview.Print(screen, "0",
-		x+plotYAxisLabelsWidth,
-		y+height-plotXAxisLabelsHeight,
-		1,
-		tview.AlignLeft, plot.axesLabelColor)
+	xAxisAreaStartX := x + plotYAxisLabelsWidth + 1
+	xAxisAreaEndX := x + width
+	xAxisAvailableWidth := xAxisAreaEndX - xAxisAreaStartX
+
+	labelMap := map[int]string{}
+	labelStartMap := map[int]int{}
+
+	maxDataPoints := 0
+	for _, d := range plot.data {
+		maxDataPoints = max(maxDataPoints, len(d))
+	}
+
+	// determine the width needed for the largest label
+	maxXAxisLabelWidth := 0
+	for _, d := range plot.data {
+		for i, _ := range d {
+			label := plot.xAxisLabelFunc(i)
+			labelMap[i] = label
+			maxXAxisLabelWidth = max(maxXAxisLabelWidth, len(label))
+		}
+	}
+
+	// determine the start position for each label, if they were
+	// to be centered below the data point.
+	// Note: not all of these labels will be printed, as they would
+	// overlap with each other
+	for i, label := range labelMap {
+		expectedLabelWidth := len(label)
+		if i == 0 {
+			expectedLabelWidth += plotXAxisLabelsGap / 2
+		} else {
+			expectedLabelWidth += plotXAxisLabelsGap
+		}
+		currentLabelStart := i - int(math.Round(float64(expectedLabelWidth)/2))
+		labelStartMap[i] = currentLabelStart
+	}
+
+	// print the labels, skipping those that would overlap,
+	// stopping when there is no more space
+	lastUsedLabelEnd := math.MinInt
+	initialOffset := xAxisAreaStartX
+	for i := 0; i < maxDataPoints; i++ {
+		labelStart := labelStartMap[i]
+		if labelStart < lastUsedLabelEnd {
+			// the label would overlap with the previous label
+			continue
+		}
+
+		rawLabel := labelMap[i]
+		labelWithGap := rawLabel
+		if i == 0 {
+			labelWithGap += strings.Repeat(gapRune, plotXAxisLabelsGap/2)
+		} else {
+			labelWithGap = strings.Repeat(gapRune, plotXAxisLabelsGap/2) + labelWithGap + strings.Repeat(gapRune, plotXAxisLabelsGap/2)
+		}
 
-	for labelX := x + plotYAxisLabelsWidth +
-		(plotXAxisLabelsGap)*plotHorizontalScale + 1; labelX < x+width-1; {
-		label := strconv.Itoa((labelX-(x+plotYAxisLabelsWidth)-1)/(plotHorizontalScale) + 1)
+		expectedLabelWidth := len(labelWithGap)
+		remainingWidth := xAxisAvailableWidth - labelStart
+		if expectedLabelWidth > remainingWidth {
+			// the label would be too long to fit in the remaining space
+			if expectedLabelWidth-1 <= remainingWidth {
+				// if we omit the last gap, it fits, so we draw that before stopping
+				labelWithoutGap := labelWithGap[:len(labelWithGap)-1]
+				plot.printXAxisLabel(screen, labelWithoutGap, initialOffset+labelStart, y+height-plotXAxisLabelsHeight)
+			}
 
-		tview.Print(screen, label, labelX, y+height-plotXAxisLabelsHeight, width, tview.AlignLeft, plot.axesLabelColor)
+			break
+		}
 
-		labelX += (len(label) + plotXAxisLabelsGap) * plotHorizontalScale
+		lastUsedLabelEnd = labelStart + expectedLabelWidth
+		plot.printXAxisLabel(screen, labelWithGap, initialOffset+labelStart, y+height-plotXAxisLabelsHeight)
```

Comment:
btw the performance of this method could probably be optimized for some cases, however it is already not even noticable in my taskmanager of choice even when updating the new demo every 50 milliseconds. So I think its safe to say that this is sufficiently fast for now.